flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultSchedulerTest.java

```diff
@@ -854,11 +841,85 @@ public void restoreStateWhenRestartingTasks() throws Exception {
 
         scheduler.updateTaskExecutionState(createFailedTaskExecutionState(attemptId));
         taskRestartExecutor.triggerScheduledTasks();
-        assertThat(masterHook.getRestoreCount(), is(equalTo(1)));
+        assertThat(masterHook.getRestoreCount()).isEqualTo(1);
+    }
+
+    @Test
+    void testTriggerCheckpointAndCompletedAfterStore() throws Exception {
```

Comment:
I don't quite understand why this test case can guard the correctness of your proposed fix approach. It seems that if the fix is removed, this test case can also be passed.

Reply:
The behavior we actually need to test should be:
1. Before `cleanupAfterCompletedCheckpoint` is completed, `checkpointFuture` should be in an incomplete state. After the `cleanupAfterCompletedCheckpoint` is completed, the `checkpointFuture` should be in the completed normally state.
2. If an exception occurs when the checkpoint is written to the `CheckpointStateHandleStore`, the `checkpointFuture` should not be completed or completed exceptionlly.